internal/namespace/reachabilitygraphbuilder.go

```diff
@@ -0,0 +1,226 @@
+package namespace
+
+import (
+	"fmt"
+
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+)
+
+type reachabilityOption int
+
+const (
+	reachabilityFull reachabilityOption = iota
+	reachabilityOptimized
+)
+
+func computeReachability(ts *NamespaceTypeSystem, relationName string, option reachabilityOption) (*core.ReachabilityGraph, error) {
+	targetRelation, ok := ts.relationMap[relationName]
+	if !ok {
+		return nil, fmt.Errorf("relation `%s` not found under type `%s` missing when computing reachability", relationName, ts.nsDef.Name)
+	}
+
+	if !ts.HasTypeInformation(relationName) && !ts.IsPermission(relationName) {
+		return nil, fmt.Errorf("relation `%s` missing type information when computing reachability", relationName)
+	}
+
+	graph := &core.ReachabilityGraph{
+		EntrypointsBySubjectType:     map[string]*core.ReachabilityEntrypoints{},
+		EntrypointsBySubjectRelation: map[string]*core.ReachabilityEntrypoints{},
+	}
+
+	usersetRewrite := targetRelation.GetUsersetRewrite()
+	if usersetRewrite != nil {
+		return graph, computeRewriteReachability(graph, usersetRewrite, core.ReachabilityEntrypoint_DIRECT_OPERATION_RESULT, targetRelation, ts, option)
+	}
+
+	// If there is no userRewrite, then we have a relation and its entrypoints will all be
+	// relation entrypoints.
+	return graph, addSubjectLinks(graph, []uint32{}, core.ReachabilityEntrypoint_DIRECT_OPERATION_RESULT, targetRelation, ts)
+}
+
+func computeRewriteReachability(graph *core.ReachabilityGraph, rewrite *core.UsersetRewrite, operationResultState core.ReachabilityEntrypoint_EntrypointResultStatus, targetRelation *core.Relation, ts *NamespaceTypeSystem, option reachabilityOption) error {
+	switch rw := rewrite.RewriteOperation.(type) {
+	case *core.UsersetRewrite_Union:
+		return computeRewriteOpReachability(rw.Union.Child, operationResultState, graph, targetRelation, ts, option)
+
+	case *core.UsersetRewrite_Intersection:
+		// If optimized mode is set, only return the first child of the intersection.
+		if option == reachabilityOptimized {
```

Comment:
I thought we would be storing the full reachability graph (since it's static per schema revision) and then pruning the branches at runtime based on heuristics.

These seems to be optimizing for calculating a pruned reachability graph rather than pruning an existing graph.

Reply:
That's how I originally had it, but pruning became a bit of a mess, so I decided to change to a simpler model where we just compute it in real time. I figure we use real time computation until it no longer scales.